src/core/server/plugins/discovery/plugins_discovery.test.ts

```diff
@@ -17,251 +17,376 @@
  * under the License.
  */
 
-import { mockPackage, mockReaddir, mockReadFile, mockStat } from './plugins_discovery.test.mocks';
-import { rawConfigServiceMock } from '../../config/raw_config_service.mock';
+import { mockPackage } from './plugins_discovery.test.mocks';
+import mockFs from 'mock-fs';
 import { loggingServiceMock } from '../../logging/logging_service.mock';
 
-import { resolve } from 'path';
 import { first, map, toArray } from 'rxjs/operators';
-
+import { resolve } from 'path';
 import { ConfigService, Env } from '../../config';
 import { getEnvOptions } from '../../config/__mocks__/env';
-import { PluginWrapper } from '../plugin';
 import { PluginsConfig, PluginsConfigType, config } from '../plugins_config';
 import { discover } from './plugins_discovery';
+import { rawConfigServiceMock } from '../../config/raw_config_service.mock';
+import { CoreContext } from '../../core_context';
+
+const KIBANA_ROOT = process.cwd();
+
+const Plugins = {
+  invalid: () => ({
+    'kibana.json': 'not-json',
+  }),
+  incomplete: () => ({
+    'kibana.json': JSON.stringify({ version: '1' }),
+  }),
+  incompatible: () => ({
+    'kibana.json': JSON.stringify({ id: 'plugin', version: '1' }),
+  }),
+  missingManifest: () => ({}),
+  valid: (id: string) => ({
+    'kibana.json': JSON.stringify({
+      id,
+      configPath: ['plugins', id],
+      version: '1',
+      kibanaVersion: '1.2.3',
+      requiredPlugins: [],
+      optionalPlugins: [],
+      server: true,
+    }),
+  }),
+};
 
-const TEST_PLUGIN_SEARCH_PATHS = {
-  nonEmptySrcPlugins: resolve(process.cwd(), 'src', 'plugins'),
-  emptyPlugins: resolve(process.cwd(), 'plugins'),
-  nonExistentKibanaExtra: resolve(process.cwd(), '..', 'kibana-extra'),
+const packageMock = {
+  branch: 'master',
+  version: '1.2.3',
+  build: {
+    distributable: true,
+    number: 1,
+    sha: '',
+  },
 };
-const TEST_EXTRA_PLUGIN_PATH = resolve(process.cwd(), 'my-extra-plugin');
-
-const logger = loggingServiceMock.create();
-
-beforeEach(() => {
-  mockReaddir.mockImplementation((path, cb) => {
-    if (path === TEST_PLUGIN_SEARCH_PATHS.nonEmptySrcPlugins) {
-      cb(null, [
-        '1',
-        '2-no-manifest',
-        '3',
-        '4-incomplete-manifest',
-        '5-invalid-manifest',
-        '6',
-        '7-non-dir',
-        '8-incompatible-manifest',
-        '9-inaccessible-dir',
-      ]);
-    } else if (path === TEST_PLUGIN_SEARCH_PATHS.nonExistentKibanaExtra) {
-      cb(new Error('ENOENT'));
-    } else {
-      cb(null, []);
-    }
+
+const manifestPath = (...pluginPath: string[]) =>
+  resolve(KIBANA_ROOT, 'src', 'plugins', ...pluginPath, 'kibana.json');
+
+describe('plugins discovery system', () => {
```

Comment:
Should we have a test that uses symlinks? I think that's probably a common use-case for 3rd party plugins. Good news is it looks like mock-fs supports symlinks.

Reply:
What do you have in mind, what would you symlink here, and to where? a nested plugin directory to an arbitrary other path on the FS?

Reply:
Main use case doesn't really have to do with nested directories, but I think it's probably common for deployments to symlink the `<kibana root>/plugins` directory for custom plugins.

Reply:
added test